### PR TITLE
Add weekday Spark function

### PR DESCRIPTION
**Pull request by acvictor**
Returns the day of the week for date as 0 = Monday, 1 = Tuesday, …, 6 = Sunday.

Addresses #8720

### CODE REVIEW

**Comment by acvictor** [comment]
@PHILO-HE can you please review?

**Comment by FelixYBW** [comment]
> @PHILO-HE can you please review?

@rui-mo

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 1]
```diff
@@ -496,6 +496,26 @@ struct DayOfYearFunction {
...
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // 0 = Monday, 1 = Tuesday, ..., 6 = Sunday
+  FOLLY_ALWAYS_INLINE int32_t getWeekday(const std::tm& time) {
+    return (time.tm_wday - 1 + 7) % 7;
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      int32_t& result,
+      const arg_type<Timestamp>& timestamp) {
```
Comment (PHILO-HE) [inline velox/functions/sparksql/DateTimeFunctions.h:510]: According to Spark code (see [link](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L460)), `weekday` function doesn't directly accept Timestamp input, but only accepts Date input. For Timestamp input, Spark will plan a cast expression to convert it to Date type. So I think we only need consider Timestamp type in Velox.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 2]
```diff
@@ -474,6 +474,50 @@ TEST_F(DateTimeFunctionsTest, dayofWeekTs) {
...
+  EXPECT_EQ(6, weekday(parseDate("2023-08-20")));
+  EXPECT_EQ(0, weekday(parseDate("2023-08-21")));
+  EXPECT_EQ(1, weekday(parseDate("2023-08-22")));
+  EXPECT_EQ(2, weekday(parseDate("2023-08-23")));
+  EXPECT_EQ(3, weekday(parseDate("2023-08-24")));
+  EXPECT_EQ(4, weekday(parseDate("2023-08-25")));
+  EXPECT_EQ(5, weekday(parseDate("2023-08-26")));
+  EXPECT_EQ(6, weekday(parseDate("2023-08-27")));
+
+  // Test cases from Spark's DateExpressionSuite.
```
Comment (PHILO-HE) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:495]: Not necessary comment. Please remove it.

File: velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp [hunk 3]
```diff
@@ -474,6 +474,50 @@ TEST_F(DateTimeFunctionsTest, dayofWeekTs) {
...
+  EXPECT_EQ(2, weekday(Timestamp(-1, 0)));
+  EXPECT_EQ(6, weekday(util::fromTimestampString("2023-08-20 20:23:00.001")));
+  EXPECT_EQ(0, weekday(util::fromTimestampString("2023-08-21 21:23:00.030")));
+  EXPECT_EQ(1, weekday(util::fromTimestampString("2023-08-22 11:23:00.100")));
+  EXPECT_EQ(2, weekday(util::fromTimestampString("2023-08-23 22:23:00.030")));
+  EXPECT_EQ(3, weekday(util::fromTimestampString("2023-08-24 15:23:00.000")));
+  EXPECT_EQ(4, weekday(util::fromTimestampString("2023-08-25 03:23:04.000")));
+  EXPECT_EQ(5, weekday(util::fromTimestampString("2023-08-26 01:03:00.300")));
+  EXPECT_EQ(6, weekday(util::fromTimestampString("2023-08-27 01:13:00.000")));
+  // Test cases from Spark's DateExpressionSuite.
```
Comment (PHILO-HE) [inline velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:514]: Not necessary comment.

File: velox/functions/sparksql/DateTimeFunctions.h (hunk 1 above)
Comment (acvictor) [reply velox/functions/sparksql/DateTimeFunctions.h:510]: I see that `dayofweek` also extends `GetDateField` but has been registered for `Date` and `Timestamp` in Velox. Should that be updated to just `Date` as well? https://github.com/facebookincubator/velox/blob/7b68a82638fa1a27ecb5cf4d3f82100481a812da/velox/functions/sparksql/Register.cpp#L304

Comment (PHILO-HE) [reply velox/functions/sparksql/DateTimeFunctions.h:510]: Yes, please go ahead to update it.

Comment (acvictor) [reply velox/functions/sparksql/DateTimeFunctions.h]: Updated both functions, thanks!

File: velox/docs/functions/spark/datetime.rst [hunk 4]
```diff
@@ -187,6 +187,13 @@ These functions support TIMESTAMP and DATE input types.
     Returns the `ISO-Week`_ of the year from x. The value ranges from ``1`` to ``53``.
     A week is considered to start on a Monday and week 1 is the first week with >3 days.
 
+.. function:: weekday(date/timestamp) -> integer
```
Comment (PHILO-HE) [inline velox/docs/functions/spark/datetime.rst:190]: Please remove timestamp. Ditto for `dayofweek`.

Comment (acvictor) [reply velox/docs/functions/spark/datetime.rst]: Done, thanks!

**Comment by acvictor** [comment]
@PHILO-HE can you please review again? Thanks!

**Comment by PHILO-HE** [comment]
@acvictor, the PR looks good to me! Please update pr description to reflect removing some useless code for dayofweek function.
@mbasmanova, could you please review further?

File: velox/docs/functions/spark/datetime.rst [hunk 5]
```diff
@@ -187,6 +187,13 @@ These functions support TIMESTAMP and DATE input types.
     Returns the `ISO-Week`_ of the year from x. The value ranges from ``1`` to ``53``.
     A week is considered to start on a Monday and week 1 is the first week with >3 days.
 
+.. function:: weekday(date) -> integer
+
+    Returns the day of the week for date (0 = Monday, 1 = Tuesday, …, 6 = Sunday).
```
Comment (mbasmanova) [inline velox/docs/functions/spark/datetime.rst:192]: Having both weekday and dayofweek functions that sound similar but return different results is confusing. Do you ever hear complains from users?

File: velox/functions/sparksql/DateTimeFunctions.h [hunk 6]
```diff
@@ -496,6 +490,20 @@ struct DayOfYearFunction {
   }
 };
 
+template <typename T>
+struct WeekdayFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // 0 = Monday, 1 = Tuesday, ..., 6 = Sunday
+  FOLLY_ALWAYS_INLINE int32_t getWeekday(const std::tm& time) {
+    return (time.tm_wday - 1 + 7) % 7;
```
Comment (mbasmanova) [inline velox/functions/sparksql/DateTimeFunctions.h:499]: nit: + 6 might be cleaner

File: velox/docs/functions/spark/datetime.rst [hunk 7]
```diff
@@ -187,6 +187,13 @@ These functions support TIMESTAMP and DATE input types.
     Returns the `ISO-Week`_ of the year from x. The value ranges from ``1`` to ``53``.
     A week is considered to start on a Monday and week 1 is the first week with >3 days.
 
+.. function:: weekday(date) -> integer
```
[the remark this answers is not in the document]
Comment (PHILO-HE) [reply velox/docs/functions/spark/datetime.rst:190]: Hi @mbasmanova, thanks for your comment!
Let me help clarify. I note in Spark `weeday` doesn't directly accept TIMESTAMP input (same behavior for `dayofweek`). For TIMESTAMP input, Spark will plan a cast expression to convert it to date type, which is confirmed by my test. So I think we only need consider date type in Velox.

File: velox/docs/functions/spark/datetime.rst [hunk 8]
```diff
@@ -64,13 +64,13 @@ These functions support TIMESTAMP and DATE input types.
 
         SELECT dayofyear('2016-04-09'); -- 100
 
-.. spark:function:: dayofweek(date/timestamp) -> integer
+.. spark:function:: dayofweek(date) -> integer
```
[the remark this answers is not in the document]
Comment (PHILO-HE) [reply velox/docs/functions/spark/datetime.rst:67]: See above comment. Thanks!

File: velox/docs/functions/spark/datetime.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/docs/functions/spark/datetime.rst:190]: @PHILO-HE Thank you for explaining. Would it be possible to create a separate PR to remove TIMESTAMP input from dayofweek and provide this explanation in the PR description?

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/docs/functions/spark/datetime.rst:190]: > Returns the day of the week for date/timestamp as 0 = Monday, 1 = Tuesday, …, 6 = Sunday.

@PHILO-HE This PR description needs updating. It says that the new function takes 2 types of arguments, but it actually takes only one.

[the remark this answers is not in the document]
Comment (acvictor) [reply velox/docs/functions/spark/datetime.rst:190]: I will raise a separate PR to remove TIMESTAMP from dayofweek.

File: velox/docs/functions/spark/datetime.rst (hunk 5 above)
Comment (acvictor) [reply velox/docs/functions/spark/datetime.rst:192]: It is confusing, I suppose different users use different functions depending on the indexing they're looking for.

File: velox/docs/functions/spark/datetime.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (acvictor) [reply velox/docs/functions/spark/datetime.rst:190]: @PHILO-HE @mbasmanova I have removed all changes related to dayofweek in this PR, will create a separate one to fix it, thanks!

File: velox/docs/functions/spark/datetime.rst [hunk 9]
```diff
@@ -64,13 +64,13 @@ These functions support TIMESTAMP and DATE input types.
...
 
-    Returns the day of the week for date/timestamp (1 = Sunday, 2 = Monday, ..., 7 = Saturday).
+    Returns the day of the week for date (1 = Sunday, 2 = Monday, ..., 7 = Saturday).
     We can use `dow` as alias for ::
```
[the remark this answers is not in the document]
Comment (acvictor) [reply velox/docs/functions/spark/datetime.rst]: I do see this https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L2808

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/docs/functions/spark/datetime.rst]: Is this a function name? Would you try running a Spark query using dow function?

[the remark this answers is not in the document]
Comment (acvictor) [reply velox/docs/functions/spark/datetime.rst]: According to https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L2808 and the REPL test below, dow maps to dayofweek and used in date_part but not directly.
```
scala> val df = Seq(("2015-04-08")).toDF("dt")
df: org.apache.spark.sql.DataFrame = [dt: string]

scala> val result = df.select(dayofweek('dt).alias("day")).collect()
result: Array[org.apache.spark.sql.Row] = Array([4])

scala> val result = df.select(expr("date_part('DOW', dt)").alias("day")).show()
+---+
|day|
+---+
|  4|
+---+

result: Unit = ()

scala> val result = df.select(expr("date_part('dayofweek', dt)").alias("day")).show()
+---+
|day|
+---+
|  4|
+---+

result: Unit = ()

scala> val result = df.select(dow('dt).alias("day")).collect()
<console>:27: error: not found: value dow
       val result = df.select(dow('dt).alias("day")).collect()
                              ^
```

File: velox/docs/functions/spark/datetime.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (acvictor) [reply velox/docs/functions/spark/datetime.rst:190]: Created a new issue https://github.com/facebookincubator/velox/issues/8736

File: velox/docs/functions/spark/datetime.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/docs/functions/spark/datetime.rst]: @acvictor Thank you for checking. Sounds like "We can use `dow` as alias for" needs to be removed.

[the remark this answers is not in the document]
Comment (acvictor) [reply velox/docs/functions/spark/datetime.rst]: Will update the issue.

**Comment by acvictor** [comment]
@mbasmanova I've removed all cleanup related to `dayofweek`. Does this PR look good to you?

**Comment by mbasmanova** [comment]
@acvictor CI is red. Would you take a look?

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Review by Yuhta — APPROVED** [review]

**Pull request closed without merging**